crates/ra_parser/src/grammar/types.rs

```diff
@@ -205,14 +205,15 @@ pub(super) fn for_binder(p: &mut Parser) {
 // type A = for<'a> fn() -> ();
 // fn foo<T>(_t: &T) where for<'a> &'a T: Iterator {}
 // fn bar<T>(_t: &T) where for<'a> &'a mut T: Iterator {}
+// fn baz<T>(_t: &T) where for<'a> <&'a T as Baz>::Foo: Iterator {}
 pub(super) fn for_type(p: &mut Parser) {
     assert!(p.at(T![for]));
     let m = p.start();
     for_binder(p);
     match p.current() {
         T![fn] | T![unsafe] | T![extern] => fn_pointer_type(p),
         T![&] => reference_type(p),
-        _ if paths::is_path_start(p) => path_type_(p, false),
+        _ if paths::is_path_start(p) || p.at(T![<]) => path_type_(p, false),
```

Comment:
Ah, `is_path_start` function is just confusing...

I think what w should do is:

* change `is_path_start` to return true on `<`
* introduce a new `is_use_path_start` which behaves like today's `is_path_start`
* audit each usage of `is_path_start` and use the appropriate variant

For example, today we parse this incorrectly as well:

```
trait C {
    const C: Self;
}

impl C for () {
    const C: () = ();
}


fn main() {
    match () {
        <()>::C => ()
    }
}
```

Reply:
@matklad please see the commit https://github.com/rust-analyzer/rust-analyzer/pull/1676/commits/59949fa40d18aba596673bf27afb5a0ddffdf6e6, I'm not entirely sure about all use cases of those `paths::is*_path_start()` functions.

Reply:
Sorry, I wasn't clear: `is_path_start` should return `true` in all the cases that it it returns `true` today, plus, additionally, on `T[<]`. `is_use_path_start` should work like today's `is_path_start`.  That way, `is_path_start() || is_at(T![<])` will be changed to `is_path_start()`.

Reply:
Got it 👍

Reply:
@matklad